src/backend/distributed/commands/multi_copy.c

```diff
@@ -475,175 +399,141 @@ CopyToExistingShards(CopyStmt *copyStatement, char *completionTag)
+		shardConnections = GetShardHashConnections(shardConnectionHash, shardId,
+												   &shardConnectionsFound);
+		if (!shardConnectionsFound)
+		{
+			bool stopOnFailure = false;
+
+			if (cacheEntry->partitionMethod == DISTRIBUTE_BY_NONE)
 			{
-				ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
-								errmsg("could not find shard for partition column "
-									   "value")));
+				stopOnFailure = true;
 			}
 
-			shardId = shardInterval->shardId;
+			/* open connections and initiate COPY on shard placements */
+			OpenCopyConnections(copyStatement, shardConnections, stopOnFailure,
+								copyOutState->binary);
 
-			MemoryContextSwitchTo(oldContext);
 
-			/* get existing connections to the shard placements, if any */
-			shardConnections = GetShardHashConnections(copyConnectionHash, shardId,
-													   &shardConnectionsFound);
-			if (!shardConnectionsFound)
+			/* send copy binary headers to shard placements */
+			if (copyOutState->binary)
 			{
-				bool stopOnFailure = false;
-
-				if (cacheEntry->partitionMethod == DISTRIBUTE_BY_NONE)
-				{
-					stopOnFailure = true;
-				}
-
-				/* open connections and initiate COPY on shard placements */
-				OpenCopyTransactions(copyStatement, shardConnections, stopOnFailure,
-									 copyOutState->binary);
-
-				/* send copy binary headers to shard placements */
-				if (copyOutState->binary)
-				{
-					SendCopyBinaryHeaders(copyOutState,
-										  shardConnections->connectionList);
-				}
+				SendCopyBinaryHeaders(copyOutState, shardId,
+									  shardConnections->connectionList);
 			}
+		}
 
-			/* replicate row to shard placements */
-			resetStringInfo(copyOutState->fe_msgbuf);
-			AppendCopyRowData(columnValues, columnNulls, tupleDescriptor,
-							  copyOutState, columnOutputFunctions);
-			SendCopyDataToAll(copyOutState->fe_msgbuf, shardConnections->connectionList);
+		/* replicate row to shard placements */
+		resetStringInfo(copyOutState->fe_msgbuf);
+		AppendCopyRowData(columnValues, columnNulls, tupleDescriptor,
+						  copyOutState, columnOutputFunctions);
+		SendCopyDataToAll(copyOutState->fe_msgbuf, shardId,
+						  shardConnections->connectionList);
 
-			processedRowCount += 1;
-		}
+		processedRowCount += 1;
+	}
 
-		connectionList = ConnectionList(copyConnectionHash);
+	shardConnectionsList = ShardConnectionList(shardConnectionHash);
+	foreach(shardConnectionsCell, shardConnectionsList)
+	{
+		ShardConnections *shardConnections = (ShardConnections *) lfirst(
+			shardConnectionsCell);
 
 		/* send copy binary footers to all shard placements */
 		if (copyOutState->binary)
 		{
-			SendCopyBinaryFooters(copyOutState, connectionList);
+			SendCopyBinaryFooters(copyOutState, shardConnections->shardId,
+								  shardConnections->connectionList);
 		}
 
-		/* all lines have been copied, stop showing line number in errors */
-		error_context_stack = errorCallback.previous;
-
```

Comment:
You've moved this to after the EndRemoteCall() below - that's what triggers marco's complaint elsewhere in this PR.  So just reset it before, and we should b egood on that end.